sui_programmability/examples/games/sources/frenemies.move

```diff
@@ -0,0 +1,387 @@
+        /// Epoch when this score was recorded
+        epoch: u64
+    }
+
+    /// Goal: validator finishes in top third by stake
+    const FRIEND: u8 = 0;
+    /// Goal: validator finishes in middle third by stake
+    const NEUTRAL: u8 = 1;
+    /// Goal: validator finishes in bottom third by stake
+    const ENEMY: u8 = 2;
+
+    /// Number of points a player receives for successfully completing an epoch goal.
+    /// Additional points are awarded according to the difficulty of the goal (i.e.,
+    /// how many places the validator moved during the epoch)
+    const GOAL_COMPLETION_POINTS: u16 = 10;
+
+    /// Attempting to call `score` on the same `Scorecard` object more than once/epoch
+    const EAlreadyUpdatedScorecardThisEpoch: u64 = 0;
+
+    fun init(ctx: &mut TxContext) {
+        let epoch = tx_context::epoch(ctx);
+        transfer::share_object(
+            Leaderboard {
+                id: object::new(ctx),
+                prev_epoch_stakes: table::new(ctx),
+                epoch,
+                start_epoch: epoch,
+            }
+        )
+    }
+
+    /// Get the current assignment for `addr`.
+    /// Always returns the same assignment for a fixed address A and epoch E.
+    fun get_assignment(addr: address, state: &SuiSystemState): Assignment {
+        let validators = validator_set::active_validators(sui_system::validators(state));
+        let len = vector::length(validators);
+
+        // derive assignment from the stake of top validator + address.
+        // this means that for a fixed address A and epoch E, the assignment is deterministic
+        let top_stake = (validator::total_stake(vector::borrow(validators, 0)) as u256);
+        let addr = address::to_u256(addr);
+        let assignment_seed = (addr ^ top_stake);
+        let validator_idx = assignment_seed % (len as u256);
+        let validator = validator::sui_address(vector::borrow(validators, (validator_idx as u64)));
+        let goal = ((assignment_seed % 3) as u8);
+        Assignment {
+            validator,
+            goal,
+        }
+    }
+
+    /// Register the transaction sender for the frenemies game by sending them a 0 scorecard with
+    /// an initial assignment
+    public entry fun register(state: &SuiSystemState, ctx: &mut TxContext) {
+        let sender = tx_context::sender(ctx);
+        transfer::transfer(
+            Scorecard {
+                id: object::new(ctx),
+                assignment: get_assignment(sender, state),
+                score: 0,
+                participation: 0,
+                epoch: tx_context::epoch(ctx),
+            },
+            sender
+        )
+    }
+
+    /// Update `scorecard` with the results from the last epoch.
+    /// Note: relies on someone calling this at least once/epoch to preserve
+    /// the record of validator scores
+    public entry fun update(
+        scorecard: &mut Scorecard, state: &SuiSystemState, leaderboard: &mut Leaderboard, ctx: &mut TxContext
```

Comment:
The leaderboard really just needs to be updated once per epoch and the system state object is solely here for that purpose. (It's also used right now for getting the assignment but we can use the stake info stored in the leaderboard instead.)

I think in general it's better to reduce the use of the shared system state object since it's also used in all the staking/delegation/tallying/gas price operations. Maybe we can have another version of the entry funs `register` and `update` that don't take in the system state obj and `assert` that `leaderboard.epoch == cur_epoch`.

Reply:
This broadly makes sense to me, but I'm worried about what will happen at the beginning of a new epoch. My expectation is that a lot of users will call `update` ASAP, but this will fail unless someone has called the function that adds the stake info to the `Leaderboard` already.

If we are willing to tweak the wave 2 epoch change code to do this, that would be one way to solve the problem, but I know we (reasonably) wanted the epoch change code to be as separate as possible from the game.

Reply:
Fair point. I also can't think of a good way of doing this without tweaking the advance epoch code, which I don't think we should do. If it's not too hard maybe we can add frenemies game operations to stress testing too, emitting them together with staking operations to see how we handle these.